### DISCUSSION TITLE
fetch와 axios의 차이를 알려주세요

### DISCUSSION THREAD

**Discussion by javascriptbiggosoo** [wanted-9th-3team/frontend-study#94] [프론트 기술 질문 공간]
.

**Comment by michoball** [comment]
fetch
es6에 생긴 js 내장 라이브러리이다. Promise기반이다. 
 자바스크립트 인터페이스를 제공하며 브라우져 환경(클라이언트 사이드)에서 사용이 가능하다. 기본적으로 http 통신으로 하고 싶은 거의 모든 것을 할 수 있다.  

fetch()메서드는 URL 주소를 첫번제 값으로 받고  http 통신을 위한 header 나 body등 다양한 옵션들을 두번제 값으로 받을 수 있다. 반환되는 값은 Promise값으로 반환된다. 

axios 에서 제공하는 기능을 직접 구현할 수 있다. 
기본 fetch 사용 코드
```javascript

const url = "https://jsonplaceholder.typicode.com/todos";
const options = {
  method: "POST",
  headers: {
    Accept: "application/json",
    "Content-Type": "application/json;charset=UTF-8",
  },
  body: JSON.stringify({
    a: 10,
    b: 20,
  }),
};
fetch(url, options)
  .then((response) => response.json())
  .then((data) => {
    console.log(data);
  });
```

axios 
 XMLHttpRequest 를 사용한 node.js와 브라우저를 위한 Promise 기반 HTTP 클라이언트이다. 
fetch에서 직접 여러 값들을 입력해야하는 반면 axios 는 보다 간편하게 사용할 수 있도록 해준다.

axios 사용예제
```javascript
const url = 'https://jsonplaceholder.typicode.com/posts'
const data = {
  a: 10,
  b: 20,
};
axios
  .post(url, data, {
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json;charset=UTF-8",
    },
  })
```
http 통신으로 받아온 data를따로 작업하지 않아도 json형식으로 변환해준다. 
body 에 넣을 data도 알아서 직렬화해준다. 

차이점
-  모던 브라우져에서 사용할거면 선호도에 따라 아무거나 사용해도 좋다. ( 외부 의존성을 높이기 싫다면 fetch가 좋고 간편하게 사용하려면 axios가 좋다.) 그러나 오래된 브라우져 환경에서도 돌아가는 앱을 만드려면 XMLHttpRequest로 통신하는 axios를 사용해야한다.

- 응답 제한 시간을 둘다 정할 수 있는데 axios 는 timeout 값만 지정하면 되는 반면에  fetch 는 AbortController interface 사용해야한다 -> 쉽지 않다.
예
``` javascript
const controller = new AbortController();
const options = {
  method: 'POST',
  signal: controller.signal,
  body: JSON.stringify({
    firstName: 'David',
    lastName: 'Pollock'
  })
};  
const promise = fetch('/login', options);
const timeoutId = setTimeout(() => controller.abort(), 4000);

promise
  .then(response => {/* handle the response */})
  .catch(error => console.error('timeout exceeded'));
```

- axios는 http interceptors를 제공해 http 요청과 응답 간의 통신을 가로채서 특정 동작을 수행할 수 있다.

**Comment by javascriptbiggosoo** [comment]
### fetch
- 장점:
내장 라이브러리: 웹 브라우저의 대부분에 기본적으로 포함되어 있어서 별도의 설치나 임포트 없이 사용할 수 있습니다.
Promise 기반: Promise를 반환하므로 then과 catch로 쉽게 체이닝 할 수 있습니다.
- 단점:
에러 핸들링: 기본적으로 HTTP 오류 상태 (예: 404, 500)에서도 실패한 Promise를 반환하지 않습니다. 오류 상태를 적절히 처리하기 위해서 추가적인 로직이 필요합니다.
기능: fetch는 기본적인 HTTP 요청 기능만 제공합니다. 타임아웃, 요청 취소 등의 고급 기능을 사용하려면 추가적인 작업이 필요합니다.
### axios
- 장점:
에러 핸들링: HTTP 오류 상태에서 자동으로 실패한 Promise를 반환합니다. 이로 인해 에러 처리가 더 간편합니다.
인터셉터: 요청 및 응답을 가로채어 변경하거나 로깅하는 등의 기능을 제공합니다.
취소 기능: 진행 중인 요청을 취소하는 기능을 내장하고 있습니다.
자동 변환: JSON 데이터를 자동으로 JavaScript 객체로 변환합니다.
브라우저 및 Node.js 지원: 브라우저뿐만 아니라 Node.js 환경에서도 동작합니다.
- 단점:
외부 라이브러리: 별도의 설치 및 임포트가 필요합니다. 프로젝트에 추가적인 종속성을 포함시킬 수 있습니다.

많은 개발자들은 axios의 사용자 친화적인 API, 에러 핸들링, 인터셉터 등의 고급 기능을 선호합니다. 따라서 복잡한 웹 애플리케이션을 개발할 때 axios를 사용하는 것이 좋을 수 있습니다.

그러나 간단한 요청을 보낼 때 또는 추가적인 라이브러리 종속성을 피하고 싶을 때는 fetch를 사용하는 것이 좋습니다.

**Comment by WooGie911** [comment]
## Fetch

- ES6부터 들어온 JavaScript 내장 라이브러리입니다.
- Promise 기반으로 만들어졌기 때문에 axios와 마찬가지로 데이터 다루기가 쉽고, 내장 라이브러리라는 장점으로 상당히 편리합니다.
- JSON 데이터 핸들링 위해 추가 절차 필요합니다. (Fetch interface 의 json() 이용하는 로직 추가하여 가능)
- Intercept Requests가 기본적으로 제공되지 않습니다. (Global Fetch Method를 Overwrite 하여 나의 인터셉터 정의 가능)

### Fetch 장점
1. 자바스크립트의 내장 라이브러리 이므로 별도로 import 할 필요가 없습니다.
2. Promise 기반으로 만들어졌기 때문에 데이터 다루기 편리합니다.
3. 내장 라이브러리이기 때문에 업데이트에 따른 에러 방지가 가능합니다.

## Fetch 단점
1. 지원하지 않는 브라우저가 존재 (IE11...)
2. 네트워크 에러 발생 시 response timeout이 없어 기다려야 합니다.
3. JSON으로 변환해주는 과정 필요합니다.
4. 상대적으로 axios에 비해 기능이 부족합니다.

## Axios

- axios는 Node.js와 브라우저를 위한 Promise API를 활용하는 HTTP 통신 라이브러리입니다.
- 비동기로 HTTP 통신을 할 수 있으며 return을 promise 객체로 해주기 때문에 response 데이터를 다루기 쉽습니다.
- 자동 JSON 데이터 변환 지원합니다.
- wide browser를 지원합니다.
- HTTP Requests 의 Intercept가 가능합니다.

### Axios 장점
1. response timeout (fetch에는 없는 기능) 처리 방법이 존재입니다.
2. Promise 기반으로 만들어졌기 때문에 데이터 다루기 편리합니다.
3. 크로스 브라우징 최적화로 브라우저 호환성(구형 브라우저 지원)이 뛰어납니다.

### Axios 단점
1. 사용을 위해 모듈 설치 필요합니다. (npm install axios)
